### PR TITLE
4 Gewinnt

### PR DESCRIPTION
**Pull request by Popaulol**

### CODE REVIEW

File: Board.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+class Board:
+    board: list[list[str]]
+    size: int
+    def __init__(self, size) -> None:
+        self.size = size
+        self.board = self.create_board(size)
+
+    def create_board(self, size) -> list[list[str]]:
+        return [["_" for _ in range(size)] for _ in range(size)]
```
Comment (stefanluth) [inline Board.py]:
```suggestion

    @staticmethod
    def create_board(size: int) -> list[list[str]]:
        return [["_" for _ in range(size)] for _ in range(size)]
```

Methods that don't use any instance/class variables can be made static. See [staticmethod](https://www.tutorialsteacher.com/python/staticmethod-decorator) :)

File: Board.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        print("|", end="")
+        for row in self.board:
+            for col in row:
+                print(col + "|", end="")
+            print("\n|", end="")
+        for i in range(1, self.size + 1):
+            print(str(i) + "|", end="")
+        print()
+
+    def is_valid_turn(self, col) -> bool:
```
Comment (stefanluth) [inline Board.py]:
```suggestion
    def is_valid_turn(self, col: int) -> bool:
```

File: Board.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        return self.board[0][col] == "_"
+
+    def add_coin_to_board(self, col, symbol) -> None:
```
Comment (stefanluth) [inline Board.py]:
```suggestion
    def add_coin_to_board(self, col: int, symbol: str) -> None:
```

File: MainGame.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    player1: Player
+    player2: Player
+    board: Board
+
+    def __init__(self) -> None:
+        self.player1 = None
+        self.player1 = None
+    
+    def get_user_name(self, player_number) -> str:
+        return input(f"Player {player_number}: Please enter your name: ")
```
Comment (stefanluth) [inline MainGame.py:15]: Add the staticmethod decorator :)

File: MainGame.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    def prepare_game(self) -> str:
+        self.board = Board(7)
+
+        self.player1 = Player(self.get_user_name(1), "X")
+        self.player2 = Player(self.get_user_name(2), "O")
+
+    def turn(self, player) -> None:
```
Comment (stefanluth) [inline MainGame.py]:
```suggestion
    def turn(self, player: Player) -> None:
```

File: MainGame.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        return True
+        
+    def round(self) -> None:
+        return self.turn(self.player1) and self.turn(self.player2)
+
+    def play(self) -> None:
+        while self.round():
+            pass
+
+    def get_col_input(self, player) -> int:
```
Comment (stefanluth) [inline MainGame.py:52]:
```suggestion
    def get_col_input(self, player: Player) -> int:
```

File: MainGame.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                col = int(input())
+            except ValueError:
+                print("That sadly isn't a valid number, please try again: ")
+                continue
+
+            if 0 < col < (self.board.size + 1):
+                return col - 1 # We subtract one, because we work with 0 indexed list, whilst the Users expect 1 indexing.
+            else:
+                print(f"This index is out of range of valid column indecies. Please enter a number between 1 and {self.board.size}")
+                continue
```
Comment (stefanluth) [inline MainGame.py:65]: The out of bounds check could be included in the `Board.is_valid_turn()` :)

File: Player.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
+class Player:
+    name: str
+    symbol: str
+    def __init__(self, name: str, symbol: str) -> None:
+        self.name = name
+        self.symbol = symbol
+
```
Comment (stefanluth) [inline Player.py:7]: Add a blank line at the end of the file :) This has technical (and historical) [reasons](https://en.wikipedia.org/wiki/End-of-file) :)

File: StatusValidator.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        for row in range(len(board) - 3):
+            if board[row][col] == symbol and board[row + 1][col - 1] == symbol and board[row + 2][col - 2] == symbol and board[row + 3][col - 3] == symbol:
+                return True
+
+    return False
+
+
+def is_draw(board):
+    return all(all(place != "_" for place in row) for row in board)
+
```
Comment (stefanluth) [inline StatusValidator.py:31]: End of file :)

**Review by stefanluth — CHANGES_REQUESTED** [review]

File: MainGame.py (hunk 7 above)
Comment (Popaulol) [reply MainGame.py:65]: Yes Definitely. Error checking in multiple different places isn't good.

**Pull request closed without merging**